### ISSUE TITLE
LoopWhichDoesNotLoopInspection should not be emitted when type `iterable`

### ISSUE THREAD

**Issue by andrew-demb** [kalessil/phpinspectionsea#1662]
Description (screenshot):
---
When I have `iterable` type, there is no way to fetch the first element without `foreach` statement, but plugin emits `LoopWhichDoesNotLoopInspection` issue when foreach with one statement used.
![image](https://user-images.githubusercontent.com/12499813/100128035-bf31bf00-2e4d-11eb-85b0-434f16a9c419.png)

My suggestion - do not emit the issue when code iterates over the `iterable` type.

Example
---
```php
<?php

interface PersonalInterface
{
    /**
     * @param string $type
     *
     * @return \stdClass[]
     */
    public function getDocumentsByType(string $type): iterable;
}

class Personal implements PersonalInterface
{
    /** @var \stdClass[] */
    private array $documents;

    public function getDocumentsByType(string $type): iterable
    {
        foreach ($this->documents as $document) {
            if ($document->type !== $type) {
                continue;
            }

            yield $document;
        }
    }
}

function extractFirstPassport(PersonalInterface $personal): ?stdClass
{
    $passports = $personal->getDocumentsByType('passport');

    foreach ($passports as $passport) {
        return $passport;
    }

    return null;
}
```

**Comment by dryabov** [comment]
What not just `return $passports->current();` instead of `foreach`?

**Comment by andrew-demb** [comment]
> What not just return $passports->current(); instead of foreach?

Because `$passports` type `iterable`. If implementation will return array, code will be broken

**Comment by ktomk** [comment]
The inspection message is right, that is a loop that does not loop. For a more pure `iterable` use `iterator_to_array($passports, false)[0]` to not loop w/ `foreach`, otherwise what @dryabov wrote looks legit  for `Generator` to me.

**Comment by andrew-demb** [comment]
`iterator_to_array` works with `\Traversable`. My contract allows returning `array` also.

See example with two implementations - return Generator and return an array
Both can be used by function with `foreach`: https://3v4l.org/ikuHp

Here added function with `iterator_to_array` usage.
Broken when implementation return `array`: https://3v4l.org/1Fd0q

I don't know any simple way to fetch the first element from `iterable` except `foreach`, so I think the plugin shouldn't emit an issue.

**Comment by ktomk** [comment]
Right, `iterator_to_array`  has no array fall-through. would have been nice thought. Maybe this:

~~~
return call_user_func(static fn($x = null) => $x, ...$passports);
~~~
https://3v4l.org/NIcGP

**Comment by andrew-demb** [comment]
Solution with unpacking (`...$passports`) will fetch all data from generator instead only the first one :(

In my opinion, the plugin should emit issues with some code smells or possible errors, when available other, more clear way to implement the same functionality.

Yes, it is the loop, that does not loop. But it is intended and other ways to implement have disadvantages in the ways how it works.

`iterable` may be used in many places and alternative with the same functionality will seem like this:
```php
function firstFromIterable(iterable $iterable)
{
    if (is_array($iterable)) {
        return current($iterable);
    }

    if ($iterable instanceof IteratorAggregate) {
        return $iterable->getIterator()->current();
    }

    return $iterable->current();
}
```
But it is very verbose.

**Comment by ktomk** [comment]
Indeed, this requires a lot of knowledge for mimic foreach() for the first on any `iterable`:

```php
// iterable is array or Traversable (Iterator|IteratorAggregate)
function firstFromIterable(iterable $iterable)
{
    if (is_array($iterable)) {
        return current($iterable); # returns false if there is no such element
    }

    // Generator special-case: Is always rewound
    if ($iterable instanceof Generator) {
        return $iterable->current();
    }

    // Traversable -> Iterator
    $iterator = ($iterable instanceof Iterator) ? $iterable : new IteratorIterator($iterable);

    $iterator->rewind();
    if (!$iterator->valid()) {
        return false; # like current() above for array
    }

    return $iterator->current();
}
```
And the `false`  return is also fishy. IMHO much better is to implement it with foreach, return `null` otherwise (or throw an exception) and then just suppress the intention and leave a comment that this is by intention.

```php
function firstFromIterable(iterable $iterable)
{
    /** @noinspection LoopWhichDoesNotLoopInspection - no-loop by intention to handle iterable */
    foreach ($iterable as $current) {
        return $current;
    }

    return null;
}
```

This is what the suppression comments are for.

And btw. no need to argue with me in great detail, I'm just a bystander here who has been running over a similar case so was interested in discussing it a bit. I'm not suggesting to change this inspection mainly for the reason that I can not imagine how it should be detected whether a non-looping loop is intended or not. I find it great that non-looping loops are detected first of all.

**Comment by andrew-demb** [comment]
Interesting detail - in related inspector class existent mention about "special cases" and false-positive.
Maybe it is missed in exception cases list? :)

https://github.com/kalessil/phpinspectionsea/blob/856da01fe264e1a5b3bfe14bf91e51438cb1499b/src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/semanticalAnalysis/loops/LoopWhichDoesNotLoopInspector.java#L57

@ktomk thank you for discussion. I apologize if I was harsh in some messages. It makes me sad, if I will should to introduce some "util" function to fetch first element, or repeat extra line with suppression in every place because "fake loop" is only one way to do this in short and effective form. 
Inspection are useful - fully disable it is not the better choice for me, of course :)

**Comment by ktomk** [comment]
Maybe it's save when the loop only contains a single return statement passing the value. This would at least cover the first element case.

**Comment by ea-inspections-team** [comment]
Thanks you for reporting and creating a PR @andrew-demb ! The PR has been merged in March and it's a good time to close the issue as well.
